### PR TITLE
Kernel Initializer issue

### PR DESCRIPTION
**Pull request by ananno**
To fix the kernel initializer issue in function `conv2d` in `lmnet/lmnet/layers.py`.

## Motivation and Context

The argument definition of `kernel_initializer` of the function `conv2d` in `lmnet/lmnet/layers/layers.py` has never been passed into `tf.layers.conv2d` function call resulting it not effective even if it is passed from the calling function. This change is necessary to pass the parameters in `tf.layers.conv2d` function by defining and passing arguments like `kernel_initializer` or `kernel_regularizer` etc.

## Description

The `kernel_initializer` argument is defined in the function argument of `conv2d` but never passed in `tf.layers.conv2d` function in the code body. This results to not passing the kernel regularizer defined or passed by the calling function. 

The fix is very simple. By removing the argument definition in `conv2d` function in `lmnet/lmnet/layers/layers.py` file. The defined argument like `kernel_initializer` or any others supported by `tf.layers.conv2d` function can be automatically passed via `*args, **kwargs`.

Details discussion on this is documented in issue [here](https://github.com/blue-oil/blueoil/issues/362).

## How has this been tested?

From the code it is obvious that the argument definition of `kernel_initializer` in `conv2d` function has never been used.

## Screenshots (if appropriate):
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by Joeper214** [comment]
@ananno Please rebase this branch from the current `blueoil/master` branch. 

So that we can only see the commits from your branch.
Currently, there are `1,004` commits which many are unrelated to this PR.

**Comment by ananno** [comment]
> @ananno Please rebase this branch from the current `blueoil/master` branch.
> 
> So that we can only see the commits you did on your branch.
> Currently, there are `1,004` commits which many are unrelated to this PR.

I've resolved the issue. Please check and confirm.
Thank you.

**Comment by Joeper214** [comment]
run test

**Comment by iizukak** [comment]
@ananno 
Thank you for fixing the bug!!
As you comment in this PR, current lmnet implementation looks not pass `kernel_initializer` to TensorFlow.
So, currently, TensorFlow use default value `None`
https://www.tensorflow.org/api_docs/python/tf/layers/conv2d

And if your code will be merged, we use `xavier_initializer` as default (in your screenshot).

I thinks make `None` as default looks better.
How do you think?

**Comment by iizukak** [comment]
@nlpng 
This PR effect many network.
It's good to be merge?

**Comment by iizukak** [comment]
I checked TensorFlow use `glorot_uniform` when `kernel_initializer` is `None`.
So there is no problem about default initializer.

**Comment by iizukak** [comment]
run test

**Comment by Joeper214** [comment]
@ananno Please fix failed tests as well.

**Comment by ananno** [comment]
@tsawada San: Can you please do Readability review on this PR.
Thank you

**Comment by ananno** [comment]
@iizukak San: You approved this PR but I think according to new policy you've to approve the ownership review. Please approve as ownership review.

Thank you

**Comment by iizukak** [comment]
Please check CI status.

**Comment by ananno** [comment]
@iizukak @tsawada San:
I though this would be a very simple fix (passing the `kernel_initializer` through `**kwargs` but as i investigate the test failure in `yolo_v2` training it expose another significant bug. 

The `kernel_initializer` parameter passing from darknet block was never passed into `conv2D` before. But when I fixed `conv2D` it turns out the parameter passing. Which means so far, all the network inherits yolov2, i.e. `lmfyolo`, didin't use `variance_scaling_initializer` but use the default normalized xavier initializer. This should cause performance difference in YoloV2 and all it's inheritor performance. 

As you already know for relu activation it was beneficial to use `variance_scaling_initializer` as kernel_initializer but the implement was wrong. Also in TF 1.15.0 the implementation was also changed.

Anyway, I've fixed both the issue and now all the tests are passed. I think it was a good finding. Please review one more time.

**Comment by tsawada** [comment]
I think @tkng san knows better so I'll leave readability review to him.

**Comment by iizukak** [comment]
@tkng 
I think we can merge this PR.

**Comment by ananno** [comment]
@tkng San:
Can you please review one more time.

Thank you

**Comment by ananno** [comment]
@iizukak San:
Can you please review one more time and if it is ok please give a OA and LGTM on this PR.

File: lmnet/lmnet/layers/layers.py [hunk 1]
```diff
@@ -50,7 +50,6 @@ def conv2d(
     strides=1,
     padding="SAME",
     activation=tf.nn.relu,
-    kernel_initializer=tf.contrib.layers.xavier_initializer(),
```
Comment (iizukak) [inline lmnet/lmnet/layers/layers.py]: @ananno Sorry, I have one question, before your change, we use `layers.xavier_initializer` and ignore `variance_scaling_initializer` by bug. 
Your fix change initializer `xavier_initializer` to `variance_scaling_initializer` and it's affect training accuracy?  🤔

Comment (ananno) [reply lmnet/lmnet/layers/layers.py]: I didn't trained with both. But theoretically it would better perform with `variance_scaling_initializer` as it works better with `ReLU` activation whereas `xavier_initializer` works better with sigmoid activation.
Ref. Xavier : http://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf
Ref. Variance Scaling : https://arxiv.org/pdf/1502.01852v1.pdf

Comment (iizukak) [reply lmnet/lmnet/layers/layers.py]: I want @tkng san's opinion about this.

Comment (ananno) [reply lmnet/lmnet/layers/layers.py]: Moreover, before we didn't pass `kernel_initializer` into the `tf.layers.conv2d` function anyway, so it would use the default value always and our selection through config was not being applied.

Comment (iizukak) [reply lmnet/lmnet/layers/layers.py]: I see.

**Comment by ananno** [comment]
@tsawada San:
Can you please review this PR for Readability Approval.

File: lmnet/lmnet/blocks.py [hunk 2]
```diff
@@ -40,7 +40,7 @@ def activation(x): return tf.nn.leaky_relu(x, alpha=0.1, name="leaky_relu")
 
         conv = conv2d("conv", inputs, filters=filters, kernel_size=kernel_size,
                       activation=None, use_bias=False, data_format=channel_data_format,
-                      kernel_initializer=tf.contrib.layers.variance_scaling_initializer(),)  # he initializer
+                      kernel_initializer=tf.variance_scaling_initializer,)  # he initializer
```
Comment (tkng) [inline lmnet/lmnet/blocks.py]: According to the reference manual, `tf.contrib.layers.variance_scaling_initializer` is he initializer (MSRA initializer), but `tf.variance_scaling_initializer` is not. (Latter doesn't have an explicit short name, Caffe's default.)

- https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/contrib/layers/variance_scaling_initializer
- https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/initializers/variance_scaling

I think it's better to keep `tf.contrib.layers.variance_scaling_initializer()`, since he initializer is major than `tf.variance_scaling_initializer`'s default setting.

Comment (ananno) [reply lmnet/lmnet/blocks.py]: I see. In that case we can keep it for now. But it will be deprecated in TF2.0, until than we can use it.
 
Screenshot from 2020-01-29 16-35-16

**Comment by ananno** [comment]
> Please revert this change, pull request should not include any cosmetic change without reason.
> 
> (I meant this line
> 
> https://github.com/blue-oil/blueoil/blob/984cd163bd26f80efb6989fea29ca94824615603/lmnet/lmnet/blocks.py#L43
> 
> )

@tkng San:
Thank you for the explanation. I've revert the cosmetic change. Please review one more time.

**Comment by tkng** [comment]
Readability Approval
Ownership Approval
